l1-contracts/contracts/bridge/L1ERC20Bridge.sol

```diff
@@ -6,122 +6,303 @@ import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+                new bytes[](0)
             );
-            l2BridgeProxyConstructorData = abi.encode(bridgeImplementationAddr, _governor, proxyInitializationParams);
+            require(bridgeProxyAddr == l2BridgeStandardAddress, "L1EB: bridge address does not match");
+            _setTxHashes(_chainId, bridgeImplTxHash, bridgeProxyTxHash);
         }
+    }
 
-        // Deploy L2 bridge proxy contract
-        l2Bridge = BridgeInitializationHelper.requestDeployTransaction(
-            zkSync,
-            _deployBridgeProxyFee,
-            l2BridgeProxyBytecodeHash,
-            l2BridgeProxyConstructorData,
-            // No factory deps are needed for L2 bridge proxy, because it is already passed in previous step
-            new bytes[](0)
+    // to avoid stack too deep error
+    function _setTxHashes(uint256 _chainId, bytes32 _bridgeImplTxHash, bytes32 _bridgeProxyTxHash) internal {
+        bridgeImplDeployOnL2TxHash[_chainId] = _bridgeImplTxHash;
+        bridgeProxyDeployOnL2TxHash[_chainId] = _bridgeProxyTxHash;
+    }
+
+    /// @dev We have to confirm that the deploy transactions succeeded.
+    function finishInitializeChain(
+        uint256 _chainId,
+        uint256 _bridgeImplTxL2BatchNumber,
+        uint256 _bridgeImplTxL2MessageIndex,
+        uint16 _bridgeImplTxL2TxNumberInBatch,
+        bytes32[] calldata _bridgeImplTxMerkleProof,
+        uint256 _bridgeProxyTxL2BatchNumber,
+        uint256 _bridgeProxyTxL2MessageIndex,
+        uint16 _bridgeProxyTxL2TxNumberInBatch,
+        bytes32[] calldata _bridgeProxyTxMerkleProof
+    ) external {
+        require(l2BridgeAddress[_chainId] == address(0), "L1EB: bridge already deployed");
+        require(bridgeImplDeployOnL2TxHash[_chainId] != 0x00, "L1EB: bridge impl tx not sent");
+
+        require(
+            bridgehub.proveL1ToL2TransactionStatus(
+                _chainId,
+                bridgeImplDeployOnL2TxHash[_chainId],
+                _bridgeImplTxL2BatchNumber,
+                _bridgeImplTxL2MessageIndex,
+                _bridgeImplTxL2TxNumberInBatch,
+                _bridgeImplTxMerkleProof,
+                TxStatus(1)
+            ),
+            "L1EB: bridge impl tx not confirmed"
+        );
+        require(
+            bridgehub.proveL1ToL2TransactionStatus(
+                _chainId,
+                bridgeProxyDeployOnL2TxHash[_chainId],
+                _bridgeProxyTxL2BatchNumber,
+                _bridgeProxyTxL2MessageIndex,
+                _bridgeProxyTxL2TxNumberInBatch,
+                _bridgeProxyTxMerkleProof,
+                TxStatus(1)
+            ),
+            "L1EB: bridge proxy tx not confirmed"
         );
+        delete bridgeImplDeployOnL2TxHash[_chainId];
+        delete bridgeProxyDeployOnL2TxHash[_chainId];
+        l2BridgeAddress[_chainId] = l2BridgeStandardAddress;
+        l2TokenBeaconAddress[_chainId] = l2TokenBeaconStandardAddress;
```

Comment:
This is a valid problem. I think probably change the startERC20BridgeInitOnChain so that it saves the future address in a mapping, and load from there instead of the l2BridgeStandard address. And then we delete that.

Reply:
Note that we actually have the following function:

```
function l2TokenAddress(address _l1Token) public view returns (address) {
```

And it is actively used in the SDK IIRC. IMHO the easiest thing is just to commit that those bytecode hashes are the same TBH cc @vladbochok WDYT? This has been an assumption for the Era bridge too, so I don't think that creating lots of complications to handle such edge cases is needed

Reply:
I think we can leave that function as a legacy function for Era, and in the new version of the SDK we can add the chainId to the function, making it work. And then my previous solution works.

Reply:
I think we can close this.